drivers/pwm/pwm_ifx_cat1.c

```diff
@@ -0,0 +1,177 @@
+
+static int ifx_cat1_pwm_init(const struct device *dev)
+{
+	struct ifx_cat1_pwm_data *data = dev->data;
+	const struct ifx_cat1_pwm_config *config = dev->config;
+	cy_en_tcpwm_status_t status;
+	int ret;
+	uint32_t addr_offset = (uint32_t)config->reg_addr - TCPWM0_BASE;
+	uint32_t clk_connection;
+
+	const cy_stc_tcpwm_pwm_config_t pwm_config = {
+		.pwmMode = CY_TCPWM_PWM_MODE_PWM,
+		.clockPrescaler = CY_TCPWM_PWM_PRESCALER_DIVBY_1,
+		.pwmAlignment = CY_TCPWM_PWM_LEFT_ALIGN,
+		.runMode = CY_TCPWM_PWM_CONTINUOUS,
+		.countInputMode = CY_TCPWM_INPUT_LEVEL,
+		.countInput = CY_TCPWM_INPUT_1,
+	};
+
+	/* Configure PWM clock */
+	Cy_SysClk_PeriphDisableDivider(config->divider_type, config->divider_sel);
+	Cy_SysClk_PeriphSetDivider(config->divider_type, config->divider_sel, config->divider_val);
+	Cy_SysClk_PeriphEnableDivider(config->divider_type, config->divider_sel);
+
+	/* This is very specific to the cyw920829m2evk_02 and may need to be modified
+	 * for other boards.
+	 */
+	if (addr_offset < sizeof(TCPWM_GRP_Type)) {
+		clk_connection =
+			PCLK_TCPWM0_CLOCK_COUNTER_EN0 + (addr_offset / sizeof(TCPWM_GRP_CNT_Type));
+	} else {
+		addr_offset -= sizeof(TCPWM_GRP_Type);
+		clk_connection = PCLK_TCPWM0_CLOCK_COUNTER_EN256 +
+				 (addr_offset / sizeof(TCPWM_GRP_CNT_Type));
+	}
+	Cy_SysClk_PeriphAssignDivider(clk_connection, config->divider_type, config->divider_sel);
+
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Configure the TCPWM to be a PWM */
+	data->pwm_num += addr_offset / sizeof(TCPWM_GRP_CNT_Type);
+	status = Cy_TCPWM_PWM_Init(PWM_REG_BASE, data->pwm_num, &pwm_config);
+	if (status != CY_TCPWM_SUCCESS) {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int ifx_cat1_pwm_set_cycles(const struct device *dev, uint32_t channel,
+				   uint32_t period_cycles, uint32_t pulse_cycles, pwm_flags_t flags)
+{
+	struct ifx_cat1_pwm_data *data = dev->data;
+	const struct ifx_cat1_pwm_config *config = dev->config;
+
+	if (!config->resolution_32_bits &&
+	    ((period_cycles > UINT16_MAX) || (pulse_cycles > UINT16_MAX))) {
+		/* 16-bit resolution */
+		if (period_cycles > UINT16_MAX) {
+			LOG_ERR("Period cycles more than 16-bits (%u)", period_cycles);
+		}
+		if (pulse_cycles > UINT16_MAX) {
+			LOG_ERR("Pulse cycles more than 16-bits (%u)", pulse_cycles);
+		}
+		return -EINVAL;
+	}
+
+	if ((period_cycles == 0) || (pulse_cycles == 0)) {
+		Cy_TCPWM_PWM_Disable(PWM_REG_BASE, data->pwm_num);
+	} else {
+		Cy_TCPWM_PWM_SetPeriod0(PWM_REG_BASE, data->pwm_num, period_cycles);
+		Cy_TCPWM_PWM_SetCompare0Val(PWM_REG_BASE, data->pwm_num, pulse_cycles);
+
+		if ((flags & PWM_POLARITY_MASK) == PWM_POLARITY_INVERTED) {
+			config->reg_addr->CTRL &= ~TCPWM_GRP_CNT_V2_CTRL_QUAD_ENCODING_MODE_Msk;
+			config->reg_addr->CTRL |= _VAL2FLD(TCPWM_GRP_CNT_V2_CTRL_QUAD_ENCODING_MODE,
+							   CY_TCPWM_PWM_INVERT_ENABLE);
+		}
+
+		/* TODO: Add 2-bit field to top 8 bits of pwm_flags_t to set this.
```

Comment:
should we remove TODO?

Reply:
This will effect applications that care about the state of the PWM pin when the PWM is disabled.  It will only matter with certain circuitry and maybe whether or not the signal is inverted.  I do think that it is worth adding, but it is outside of the scope of this first release.